src/with_clause_parser.c

```diff
@@ -0,0 +1,153 @@
+ *        in = foo.foo_para
+ *    not_in = bar.bar_param, baz.baz_param
+ */
+void
+ts_with_clause_filter(const List *def_elems, const char *namespace, List **within_namespace, List **not_within_namespace)
+{
+	ListCell   *cell;
+
+	foreach(cell, def_elems)
+	{
+		DefElem    *def = (DefElem *) lfirst(cell);
+
+		if (def->defnamespace != NULL && pg_strcasecmp(def->defnamespace, namespace) == 0)
+		{
+			if (within_namespace != NULL)
+				*within_namespace = lappend(*within_namespace, def);
+		}
+		else if (not_within_namespace != NULL)
+		{
+			*not_within_namespace = lappend(*not_within_namespace, def);
+		}
+	}
+}
+
+/*
+ * Deserialize and apply the values in a WITH clause based on the on_arg table.
+ *
+ * This function will go through every element in def_elems and search for a
+ * corresponding argument in args, if one is found it will attemt to deserialize
+ * the argument, using that table elements deserialize function, then apply it
+ * to state.
+ *
+ * This is used to turn the list into a form more useful for our internal
+ * functions
+ */
+void
+ts_with_clauses_apply(const List *def_elems, ts_with_clause_on_arg on_arg, const WithClauseArg *args, Size nargs, void *state)
+{
+	ListCell   *cell;
+
+	foreach(cell, def_elems)
+	{
+		DefElem    *def = (DefElem *) lfirst(cell);
+		const char *value;
+		Size		i;
+		bool		argument_recognized = false;
+		bool		argument_valid = false;
+
+		for (i = 0; i < nargs; i++)
+		{
+			if (pg_strcasecmp(def->defname, args[i].arg_name) == 0)
+			{
+				Datum		val;
+
+				argument_recognized = true;
+
+				if (def->arg != NULL)
+					value = defGetString(def);
+				else
+					value = "true";
+
+				if (args[i].deserializer != NULL)
+					val = args[i].deserializer(def, value, state);
+				else
+					val = CStringGetDatum(value);
+
+				argument_valid = (*on_arg) (i, def->defname, val, state);
+				break;
+			}
+		}
+		if (!argument_recognized)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("unrecognized parameter \"%s.%s\"",
+							def->defnamespace, def->defname)));
+
+		if (!argument_valid)
+			ereport(ERROR,
+					(errcode(ERRCODE_AMBIGUOUS_PARAMETER),
+					 errmsg("invalid or duplicate parameter \"%s.%s\"",
+							def->defnamespace, def->defname)));
+	}
+}
+
+Datum
+ts_with_clause_deserialize_unimplemented(const DefElem *element, const char *value, void *state)
+{
+	elog(ERROR, "argument \"%s.%s\" not implemented", element->defnamespace, element->defname);
+}
+
+Datum
+ts_with_clause_deserialize_bool(const DefElem *element, const char *value, void *state)
+{
+	if (pg_strcasecmp(value, "true") == 0 || pg_strcasecmp(value, "on") == 0)
+		return BoolGetDatum(true);
+	else if (pg_strcasecmp(value, "false") == 0 || pg_strcasecmp(value, "off") == 0)
+		return BoolGetDatum(false);
+	else
+		TS_WITH_DESERIALIZE_ERROR(element, value, "BOOLEAN");
+}
+
+Datum
+ts_with_clause_deserialize_int32(const DefElem *element, const char *value, void *state)
+{
+	PG_TRY();
```

Comment:
can we use the `int4in` pg function?

Reply:
(if you meant "use it directly instead of making a `ts_with_clause_deserialize_int32`", these functions are really here for error reporting; it gives a much nicer `invalid value for hypertable.foo 'bar'` instead of only `invalid input syntax for integer: "bar"`)